access/slack/config.go

```diff
@@ -34,7 +34,8 @@ import (
 // Config stores the full configuration for the teleport-slack plugin to run.
 type Config struct {
 	common.BaseConfig
-	Slack common.GenericAPIConfig
+	Slack               common.GenericAPIConfig
+	AccessTokenProvider auth.AccessTokenProvider
```

Comment:
Currently `Config` can be parsed from a `toml` file for self hosted plugins, can we modify this so that `AccessTokenProvider` can also be parsed from toml?

```
type Config struct {
	Slack SlackConfig
}

type SlackConfig struct {
	common.GenericAPIConfig
	ClientID     string
	ClientSecret string
}
```

If these new config options will be used for other plugins as well as a common OAuth2 flow, we should add those files to `common.GenericAPIConfig` instead.

In addition to making this usable to self hosted, this removes the need for the cloud use case to call `NewAuthorizer` itself, so we can remove that exported function.

We should also update the example toml file to include this feature option.

```
# Example slack plugin configuration TOML file

[teleport]
# Teleport Auth/Proxy Server address.
# addr = "example.com:3025"
#
# Should be port 3025 for Auth Server and 3080 or 443 for Proxy.
# For Teleport Cloud, should be in the form "your-account.teleport.sh:443".

# Credentials generated with `tctl auth sign`.
#
# When using --format=file:
# identity = "/var/lib/teleport/plugins/slack/auth_id"    # Identity file
#
# When using --format=tls:
# client_key = "/var/lib/teleport/plugins/slack/auth.key" # Teleport TLS secret key
# client_crt = "/var/lib/teleport/plugins/slack/auth.crt" # Teleport TLS certificate
# root_cas = "/var/lib/teleport/plugins/slack/auth.cas"   # Teleport CA certs

[slack]
# Slack Bot OAuth token
token = "xoxb-11xx"

# description
client_id = "example"
client_secret = "example"

[role_to_recipients]
# Map roles to recipients.
#
# Provide slack user_email/channel recipients for access requests for specific roles. 
# role.suggested_reviewers will automatically be treated as additional email recipients.
# "*" must be provided to match non-specified roles.
#
# "dev" = "devs-slack-channel"
# "*" = ["admin@email.com", "admin-slack-channel"]

[log]
output = "stderr" # Logger output. Could be "stdout", "stderr" or "/var/lib/teleport/slack.log"
severity = "INFO" # Logger severity. Could be "INFO", "ERROR", "DEBUG" or "WARN".
```

Reply:
We weren't exactly planning to extend this "proper" OAuth2 support to self-hosted version just yet. This implementation is mostly to scratch the Cloud's itch :slightly_smiling_face: 

One problem is that the Slack plugin itself does not implement the initial "exchange" flow (this is planned to be implemented by Proxy/Auth in Cloud). We would need to add a HTTP server to receive the auth callback (in a similar vein to `tsh` browser login flow).

In theory, we could support this, but for now the users would need to write their own shim to do the initial auth/exchange, then configure client-id/client-secret/token. Not sure if it's worth half-baking such solution into self-hosted version. /cc @jimbishopp